torch/_inductor/mkldnn.py

```diff
@@ -283,31 +282,21 @@ def pack_module(gm: torch.fx.GraphModule):
                     and not torch.ops.mkldnn._is_mkldnn_bf16_supported()
                 ):
                     continue
-                if dynamo_config.dynamic_shapes:
-                    computation_node_input_size = None
-                    # Conv2d and ConvTranspose2d weight format are dependent on input size,
-                    # but ShapeProp may be failed to get the input size, so we skip them.
-                    if not (
-                        type(cur_module) in [torch.nn.Linear]
-                        and cur_module.weight.dtype == torch.bfloat16
-                    ):
+                computation_node_input_size = (
+                    node.args[0].meta.get("tensor_meta").shape
+                )
+                if any(size == 0 for size in computation_node_input_size):
+                    continue
+                if type(cur_module) in [torch.nn.Linear]:
+                    # for fp32 linear, only packed when has mkl.
+                    if (
+                        cur_module.weight.dtype == torch.float32
+                        and (not torch._C.has_mkl)
+                    ) or len(computation_node_input_size) < 2:
                         continue
                 else:
-                    computation_node_input_size = (
-                        node.args[0].meta.get("tensor_meta").shape
-                    )
-                    if any(size == 0 for size in computation_node_input_size):
+                    if len(computation_node_input_size) != 4:
```

Comment:
We need known input sizes to pack the weights for convolution and fp32 linear. Would `computation_node_input_size` be symbolic shape when the dynamic shape is on? If so, this would break the weight repacking. I think we should either avoid weight prepacking just as the original code does or use the size hints and add guards. The latter would incur repacking cost at runtime. cc @XiaobingSuper

Reply:
I think the way I will resolve this short term is disable the optimization if it is not statically known what the input size is (instead of using dynamic_shapes to do it).

Reply:
We can also force a guard when you find that this optimization is possible; this is a policy decision and I don't know how important this optimization is. Generally we don't do this (except for when a user explicitly asks for an optimization, like mode="reduce-overhead")

Reply:
> I think the way I will resolve this short term is disable the optimization if it is not statically known what the input size is (instead of using dynamic_shapes to do it).
> We can also force a guard when you find that this optimization is possible; this is a policy decision and I don't know how important this optimization is. Generally we don't do this (except for when a user explicitly asks for an optimization, like mode="reduce-overhead")

How about we disable the weight prepacking only when `computation_node_input_size` is symbolic? It is a big optimization without which, we would lose ~20% performance...